drop-transfer/src/ws/events.rs

```diff
@@ -44,28 +36,70 @@ pub struct FileEventTx<T: Transfer> {
     file_id: FileId,
 }
 
-pub struct FileEventTxFactory {
+pub struct EventTxFactory {
     events: UnboundedSender<(Event, SystemTime)>,
     moose: Arc<dyn Moose>,
 }
 
-impl FileEventTxFactory {
+pub struct TransferEventTx<T: Transfer> {
+    inner: Mutex<TransferEventTxInner>,
+    pub xfer: Arc<T>,
+}
+
+pub type IncomingTransferEventTx = TransferEventTx<IncomingTransfer>;
+pub type OutgoingTransferEventTx = TransferEventTx<OutgoingTransfer>;
+
+enum TransferState {
+    Ongoing,
+    Terminated,
+}
+
+struct TransferEventTxInner {
+    tx: UnboundedSender<(Event, SystemTime)>,
+    moose: Arc<dyn Moose>,
+    state: TransferState,
+}
+
+trait EventTx {
+    fn emit(&self, event: Event);
+}
+
+impl EventTx for UnboundedSender<(Event, SystemTime)> {
+    fn emit(&self, event: Event) {
+        // Sometimes on shutdown it can error out. It's better not to handle this error
+        // at all
+        let _ = self.send((event, SystemTime::now()));
```

Comment:
Can we at least log it as a warn? This way if we would encounter a weird situation and would have logs it might provide good insight

Reply:
Um, there's no way to pass logger here sadly. We do have logs on the receiving end so that needs to suffice IMHO.
Previously it was an `.expect()` on this and I saw that it sometimes panicked in the tests. I think it was because the event callback was destroyed by Python too early, so I consider failures here to happen only on weird misuse of the event callback.

Reply:
Fair enough